cmd/container-disk-v2alpha/main.c

```diff
@@ -0,0 +1,142 @@
+            printf("socket %s does not exist anymore\n", copy_path);
+            exit(0);
+        }
+    }
+}
+
+int main(int argc, char **argv) {
+    char *copy_path = NULL;
+    char *copy_path_dir;
+    bool health_check = false;
+
+    int c;
+    while(1) {
+        static struct option long_options[] = {
+            /* These options don’t set a flag.
+                We distinguish them by their indices. */
+            {"copy-path",    required_argument, 0, 'c'},
+            {"health-check", no_argument,       0, 'p'},
+            {0, 0, 0, 0}
+        };
+        /* getopt_long stores the option index here. */
+        int option_index = 0;
+        c = getopt_long(argc, argv, "c:p", long_options, &option_index);
+
+        /* Detect the end of the options. */
+        if (c == -1) {
+            break;
+        }
+
+        switch (c) {
+            case 'c':
+                copy_path = optarg;
+                copy_path_dir = strdup(copy_path);
+                dirname(copy_path_dir);
+                break;
+
+            case 'p':
+                health_check = true;
+                break;
+
+            case '?':
+                exit(1);
+            default:
+                abort();
+        }
+    }
+
+    struct stat st = {0};
+    if (health_check) {
+        if (stat(READINESS_PROBE_FILE, &st) == -1) {
+            printf("readiness probe %s does not exist\n", READINESS_PROBE_FILE);
+            exit(1);
+        } else {
+            exit(0);
+        }
+    }
+
+    if (stat(copy_path_dir, &st) == -1) {
+        if (mkdir(copy_path_dir, 0777) != 0) {
+            printf("failed to create disk directory %s\n", copy_path_dir);
+            exit(1);
+        }
+    }
+
+    struct sockaddr_un address;
+    /*
+    * For portability clear the whole structure, since some
+    * implementations have additional (nonstandard) fields in
+    * the structure.
+    */
+    memset(&address, 0, sizeof(struct sockaddr_un));
+    address.sun_family = AF_UNIX;
+    strcat(copy_path, ".sock");
+    strcpy(address.sun_path, copy_path);
+
+    int fd = socket(AF_UNIX, SOCK_STREAM, 0);
+    if (fd == -1) {
+        printf("failed to create socket on %s\n", copy_path);
+        exit(1);
+    }
+
+    if (bind(fd, (struct sockaddr*)(&address), sizeof(struct sockaddr_un)) == -1) {
+        printf("failed to bind socket %s\n", copy_path);
+        exit(1);
+    }
+
+    if (listen(fd, LISTEN_BACKLOG) == -1) {
+        printf("failed to listen socket %s\n", copy_path);
+        exit(1);
+    }
+
+    // Create readiness probe
+    FILE *probe;
+    probe = fopen(READINESS_PROBE_FILE, "w");
+    if (probe == NULL) {
+        printf("failed to create readiness probe\n");
+        exit(1);
+    }
+    fclose(probe);
+
+    pthread_t socket_check_thread;
+    pthread_create(&socket_check_thread, NULL, socket_check, (void *)copy_path);
+
+    int data_socket;
+    for (;;) {
+        data_socket = accept(fd, NULL, NULL);
```

Comment:
I  don't think that we have to `accept` connections. I would have to try it out, but `listen` above should already be sufficient. If that' the case, we could just live with an interrupt handler instead of a thread.

Reply:
@cynepco3hahue did not fully check the signal handling, but the VMIs could be started and stopped successfully without the need to accept connections on the socket:

```diff
diff --git a/cmd/container-disk-v2alpha/main.c b/cmd/container-disk-v2alpha/main.c
index b0fd3d855..8f24f82d9 100644
--- a/cmd/container-disk-v2alpha/main.c
+++ b/cmd/container-disk-v2alpha/main.c
@@ -1,7 +1,6 @@
 #include <ctype.h>
 #include <getopt.h>
 #include <libgen.h>
-#include <pthread.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
@@ -11,10 +10,18 @@
 #include <sys/types.h>
 #include <sys/un.h>
 #include <unistd.h>
+#include <signal.h>
 
 #define LISTEN_BACKLOG 50
 #define READINESS_PROBE_FILE "/healthy"
 
+char *copy_path = NULL;
+
+void sig_handler(int signo) {
+    unlink(copy_path);
+    exit(0);
+}
+
 static void *socket_check(void *arg) {
     struct stat st = {0};
     char *copy_path;
@@ -30,7 +37,6 @@ static void *socket_check(void *arg) {
 }
 
 int main(int argc, char **argv) {
-    char *copy_path = NULL;
     char *copy_path_dir;
     bool health_check = false;
 
@@ -98,6 +104,11 @@ int main(int argc, char **argv) {
     strcat(copy_path, ".sock");
     strcpy(address.sun_path, copy_path);
 
+   if (signal(SIGTERM, sig_handler) == SIG_ERR) {
+     printf("failed to register SIGTERM callback");
+     exit(1);
+   }
+
     int fd = socket(AF_UNIX, SOCK_STREAM, 0);
     if (fd == -1) {
         printf("failed to create socket on %s\n", copy_path);
@@ -123,20 +134,5 @@ int main(int argc, char **argv) {
     }
     fclose(probe);
 
-    pthread_t socket_check_thread;
-    pthread_create(&socket_check_thread, NULL, socket_check, (void *)copy_path);
-
-    int data_socket;
-    for (;;) {
-        data_socket = accept(fd, NULL, NULL);
-        if (data_socket == -1) {
-            if (stat(copy_path, &st) == -1) {
-                printf("socket %s does not exist anymore\n", copy_path);
-                exit(0);
-            }
-            printf("failed to accept connection\n");
-            unlink(copy_path);
-            exit(1);
-        }
-    }
+    socket_check((void *)copy_path);
 }
```

Should make the binary smaller and simpler, since no threads are required.

Reply:
I removed pthread, now let's see if CI likes it:)